rfc/rfc-0000-input-output-correlation.md

```diff
@@ -0,0 +1,217 @@
+
+* Resource is healthy, as per [Allow Resources to Report Status](https://github.com/vmware-tanzu/cartographer/pull/738)
+* Resource `observedGeneration == generation`
+
+If these assertions cannot be made, for example, if the resource does not implement the `observedGeneration` pattern,
+then Cartographer stops processing the resource and flags an error on it. Great care must also be taken that `healthy`
+status rules specified by the template author corresponds to the situation where the output is indeed a result of the
+spec with the current `observedGeneration`, otherwise outputs may not truly be the result of an input.
+
+Additionally, relying on the spec to make assertions imposes a further restriction to avoid stalling artifact
+propagation: Cartographer must not stamp out any further changes to the resource until it has finished processing the
+current spec completely. That is, it must have reached either a Healthy or Unhealthy status (as per
+[Allow Resources to Report Status](https://github.com/vmware-tanzu/cartographer/pull/738)).
+
+# Migration
+
+[migration]: #migration
+
+All templates must have `outputMatches` added to them.
+
+Any template which requires `outputMatches` against the resource spec must also be configured to identify success,
+failure and unknown states - which map to Healthy, Unhealthy and Unknown if the mechanism to be relied upon is
+[Allow Resources to Report Status](https://github.com/vmware-tanzu/cartographer/pull/738).
+
+# Drawbacks
+
+[drawbacks]: #drawbacks
+
+Making assertions against resources which require matching against the spec introduces a few downsides:
+
+* Holding spec updates is another bottleneck which can further slow processing;
+* Resources must implement `observedGeneration` pattern;
+* It must be possible to determine if the resources are still processing (Unknown state, as per
+  [Allow Resources to Report Status](https://github.com/vmware-tanzu/cartographer/pull/738)).
+* Great responsibility is placed on template authors to write correct `observedMatch` and health determination rules, as
+  they can otherwise be misleading, and cause propagation of artifacts which are not truly correlated, or worse:
+  unwanted artifacts.
+
+# Prior Art
+
+[prior-art]: #prior-art
+
+[Draft RFC Read Resources Only When In Success State](https://github.com/vmware-tanzu/cartographer/pull/556)
+
+[Allow Resources to Report Status](https://github.com/vmware-tanzu/cartographer/pull/738)
+
+# Unresolved Questions
+
+[unresolved-questions]: #unresolved-questions
+
+It is unclear how to solve for resources which require matching against their spec, but do not conform to the
+`observedGeneration` pattern. The same can be said for resources which do not expose sufficient detail to write proper
+health determination rules.
+
+# Spec. Changes
+
+[spec-changes]: #spec-changes
+
+All templates will now have a mandatory `observedMatch` field, which must be an array containing at least one element
```

Comment:
Making this optional, rather than required, will mean that Carto will still be able to work with all resources. It will also allow end users to determine if they want to tradeoff performance for artifact tracing.